### ISSUE TITLE
Value clipping for PPO loss

### ISSUE THREAD

**Issue by wbinventor** [pytorch/rl#1977]
## Motivation

It would nice to add a new option to compute the clipped value loss as used in OpenAI's Baselines for PPO:

https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/ppo2/model.py#L66-L75

Currently, the `PPOLoss().loss_critic` method calls the `torchrl.objectives.utils.distance_loss` function which supports the `"l1"`, `"l2"` and `"smooth_l1"` loss functions. Perhaps this new clipped value loss function could be implemented as a new loss type within `distance_loss`.

The clipping fraction is also commonly reported as a metric by OpenAI's Baselines and this could be useful to report from `PPOLoss` for the clipped value loss, as well as for `ClipPPOLoss` for the `loss_objective` loss term.

**Comment by vmoens** [comment]
That makes sense!

@albertbou92 Thoughts? Do you want to take care of this one?

**Comment by albertbou92** [comment]
Yes I also thought in the past about adding the PPO value clipping. 
Not sure yet whether the logic should go in PPO or in `distance_loss`. My first thought is in PPO, since I have only seen it used in PPO, but I guess could potentially be used in other methods.
In any case,  I can definitely start a PR to implement it.

**Comment by vmoens** [comment]
@matteobettini we could raise a warning if the value requires grad in that case? 

In a way what's great about a good lib is that it allows you to do loads of stuff but also tells you when you're off the road

**Comment by albertbou92** [comment]
Here is the code generally used:

```
# Compute normal values loss
value_loss = (new_v - return).pow(2)

if self.use_clipped_value_loss:
    value_pred_clipped = old_v + (new_v - old_v).clamp(-self.clip_param, self.clip_param)
    value_loss_clipped = (value_pred_clipped - return).pow(2)
    
    # Chose the most pessimistic value prediction between clipped and non-clipped
    # so the updates are smaller
    value_loss = torch.max(value_loss, value_loss_clipped)

value_loss = 0.5 * value_loss.mean()
```
where, 
`old_v` : value predicted during collection, without grads
`new_v`: values predicted during loss computation, with grads

So thinking in terms of PPO, the `value_loss` and `value_pred_clipped` will be the same in the first batch update.
In the next iterations, we are simply mitigate excessively large updates to the value function.

We are clipping a tensor that has gradients, but this is similar to the PPO actor loss that clips the` ratio = torch.exp(new_logp - old_logp)`. 

This make sense for PPO that reuses the data multiple times. Not so sure it make sense for other on-policy methods.
Also, this requires having the original `value_pred` computed by a relatively recent policy, so not sure if it makes sense for off-policy methods.

In summary, at this point my opinion is that we should implement just for PPO. WDYT? should we make it more general?

**Comment by wbinventor** [comment]
I think your rationale for only supporting on-policy methods makes sense 👍

**Comment by albertbou92** [comment]
This PR adds the clipping for PPO --> https://github.com/pytorch/rl/pull/2005

As mentioned, we could do something similar for other On-Policy Losses.  However, the PPO value clipping generally reuses the clipping parameter of PPO.  To add it to other losses we would need to define a new parameter for the clipping. Do we want to do that? Maybe for A2C and Reinforce I would only add it if users ask for it, as it is unclear if the clipping will be necessary/useful there.

**Comment by wbinventor** [comment]
It may be useful to define a separate parameter for this value clipping threshold. It can be useful to tune this different hyperparameter separate from that for the PPO loss.

**Comment by vmoens** [comment]
> It may be useful to define a separate parameter for this value clipping threshold. It can be useful to tune this different hyperparameter separate from that for the PPO loss.

This is the plan! https://github.com/pytorch/rl/pull/2005#discussion_r1518553859

**Comment by wbinventor** [comment]
Also, it is common and useful to log the "clip fraction" separately for both the PPO loss and the clipped value loss. Perhaps these fractions could be returned as "metrics" similar to how the `"entropy"` is returned as an item in the loss `TensorDict`?

**Comment by albertbou92** [comment]
We could add them! You mean returning them after the clipping then right? 

```
clip_fraction_policy = (new_log_prob / current_log_prob).clamp(1 - clip_epsilon, 1 + clip_epsilon)
clip_fraction_policy = (new_value_prediction / old_value_prediction).clamp(1 - clip_epsilon, 1 + clip_epsilon)
```

**Comment by wbinventor** [comment]
That's right, I think that code implementation would be great to include 👍

**Comment by vmoens** [comment]
Closing as per #2005

**Comment by wbinventor** [comment]
@albertbou92 @vmoens thanks adding the clipped value loss! 

Just realized that I mis-spoke regarding the clip fraction "metric". What I've seen in Stable Baselines and elsewhere is actually the fraction of the training data that triggered the clipped objective. In the case of the clipped PPO loss, this is a "clip fraction" metric separately reported for the clipped PPO objective and also for the clipped value loss. The `value_clip_fraction` that is currently implemented is different from this. Perhaps I will open a new feature request for this, which should be much simpler to add that the clipped value loss itself.

**Comment by vmoens** [comment]
I think the fraction is merely the percentage of values that have been clipped right?
If so it's what I was suggesting in the PR.

**Comment by wbinventor** [comment]
@albertbou92 the second formulation is indeed what I'm referring to, though I think this may be a clearer version of it.

```python
clip_fraction_value = (new_value_prediction - old_value_prediction).abs() > clip_value
```

> I think the fraction is merely the percentage of values that have been clipped right?

@vmoens yes, that's correct.
